src/freenas/usr/local/bin/truenas-initrd.py

```diff
@@ -0,0 +1,159 @@
+    new_config = "\n".join(lines) + "\n"
+    if new_config != original_config:
+        with open(zfs_config_path, "w") as f:
+            f.write(new_config)
+        return True
+    return False
+
+
+def dict_factory(cursor, row):
+    d = {}
+    for idx, col in enumerate(cursor.description):
+        d[col[0]] = row[idx]
+    return d
+
+
+def get_current_gpu_pci_ids(root):
+    from middlewared.plugins.config import FREENAS_DATABASE
+    from middlewared.utils.gpu import get_gpus
+    conn = sqlite3.connect(os.path.join(root, FREENAS_DATABASE))
+    conn.row_factory = dict_factory
+    c = conn.cursor()
+    c.execute("SELECT * FROM system_advanced")
+    adv_config = {k.replace("adv_", ""): v for k, v in c.fetchone().items()}
+    to_isolate = [gpu for gpu in get_gpus() if gpu["addr"]["pci_slot"] in adv_config["isolated_gpu_pci_ids"]]
+    return [dev["pci_id"] for gpu in to_isolate for dev in gpu["devices"]]
+
+
+def update_module_files(root, config):
+    def get_path(p):
+        return os.path.join(root, p)
+
+    pci_ids = config['pci_ids']
+    for path in map(
+        get_path, [
+            'etc/initramfs-tools/modules',
+            'etc/modules',
+            'etc/modprobe.d/kvm.conf',
+            'etc/modprobe.d/nvidia.conf',
+            'etc/modprobe.d/vfio.conf',
+        ]
+    ):
+        with contextlib.suppress(OSError):
+            os.unlink(path)
+    if not pci_ids:
+        return
+
+    os.makedirs(get_path('etc/initramfs-tools'), exist_ok=True)
+    os.makedirs(get_path('etc/modprobe.d'), exist_ok=True)
+
+    for path in map(get_path, ['etc/initramfs-tools/modules', 'etc/modules']):
+        with open(path, 'w') as f:
+            f.write(f'vfio\nvfio_iommu_type1\nvfio_virqfd\nvfio_pci ids={",".join(pci_ids)}\n')
+
+    with open(get_path('etc/modprobe.d/kvm.conf'), 'w') as f:
+        f.write('options kvm ignore_msrs=1\n')
+
+    with open(get_path('etc/modprobe.d/nvidia.conf'), 'w') as f:
+        f.write('softdep nouveau pre: vfio-pci\nsoftdep nvidia pre: vfio-pci\nsoftdep nvidia* pre: vfio-pci\n')
+
+    with open(get_path('etc/modprobe.d/vfio.conf'), 'w') as f:
+        f.write(f'options vfio-pci ids={",".join(pci_ids)}\n')
+
+
+def update_initramfs_config(root):
+    initramfs_config_path = os.path.join(root, "boot/initramfs_config.json")
+    initramfs_config = {
+        "pci_ids": get_current_gpu_pci_ids(root),
+    }
+    original_config = None
+    if os.path.exists(initramfs_config_path):
+        with open(initramfs_config_path, "r") as f:
+            original_config = json.loads(f.read())
+
+    if initramfs_config != original_config:
+        with open(initramfs_config_path, "w") as f:
+            f.write(json.dumps(initramfs_config))
+
+        update_module_files(root, initramfs_config)
+        return True
+
+    return False
+
+
+if __name__ == "__main__":
+    root = sys.argv[1]
+    if root != "/":
+        sys.path.append(os.path.join(root, "usr/lib/python3/dist-packages/middlewared"))
```

Comment:
Should it be `sys.path.insert(0, ` to give it a priority if we are running from some root where we already have middlewared?